golem-component-service-base/src/service/component.rs

```diff
@@ -404,106 +650,123 @@ where
+        )
+        .await
+    }
 
-        let new_type_registry = FunctionTypeRegistry::from_export_metadata(&metadata.exports);
+    async fn update_internal(
+        &self,
+        component_id: &ComponentId,
+        data: Vec<u8>,
+        component_type: Option<ComponentType>,
+        files: Option<Vec<InitialComponentFile>>,
+        namespace: &Namespace,
+    ) -> Result<Component<Namespace>, ComponentError> {
+        info!(namespace = %namespace, "Update component");
 
-        if let Some(constraints) = constraints {
-            let conflicts =
-                Self::find_component_metadata_conflicts(&constraints, &new_type_registry);
-            if !conflicts.is_empty() {
-                return Err(ComponentError::ComponentConstraintConflictError(conflicts));
+        for file in files.iter().flatten() {
+            let exists = self
+                .initial_component_files_service
+                .exists(&file.key)
+                .await
+                .map_err(|e| {
+                    ComponentError::initial_component_file_upload_error(
+                        "Error checking if file exists",
+                        e,
+                    )
+                })?;
+
+            if !exists {
+                return Err(ComponentError::initial_component_file_not_found(
+                    &file.path, &file.key,
+                ));
             }
         }
 
-        let next_component = self
-            .component_repo
-            .get_latest_version(&component_id.0)
-            .await?
-            .filter(|c| c.namespace == namespace.to_string())
-            .ok_or(ComponentError::UnknownComponentId(component_id.clone()))
-            .and_then(|c| {
-                c.try_into()
-                    .map_err(|e| ComponentError::conversion_error("record", e))
-            })
-            .map(Component::next_version)?;
-
-        info!(namespace = %namespace, "Uploaded component - exports {:?}", metadata.exports);
-
-        let component_size: u64 = data.len().try_into().map_err(|e: TryFromIntError| {
-            ComponentError::conversion_error("data length", e.to_string())
-        })?;
-
-        tokio::try_join!(
-            self.upload_user_component(&next_component.versioned_component_id, data.clone()),
-            self.upload_protected_component(&next_component.versioned_component_id, data)
-        )?;
-
-        let component = Component {
-            component_size,
-            metadata,
-            created_at,
-            component_type: component_type.unwrap_or(next_component.component_type),
-            ..next_component
-        };
-        let record = component
-            .clone()
-            .try_into()
-            .map_err(|e| ComponentError::conversion_error("record", e))?;
-
-        self.component_repo.create(&record).await?;
-
-        self.component_compilation
-            .enqueue_compilation(component_id, component.versioned_component_id.version)
-            .await;
-
```

Comment:
Ok I got the answer for my own question in PR, if the constraint check is not duplicated. Thanks